Kotlin-compose/build/compose-benchmarks-benchmarks.uninstantiated.mjs

```diff
@@ -0,0 +1,443 @@
+
+export async function instantiate(imports={}, runInitializer=true) {
+    const cachedJsObjects = new WeakMap();
+    // ref must be non-null
+    function getCachedJsObject(ref, ifNotCached) {
+        if (typeof ref !== 'object' && typeof ref !== 'function') return ifNotCached;
+        const cached = cachedJsObjects.get(ref);
+        if (cached !== void 0) return cached;
+        cachedJsObjects.set(ref, ifNotCached);
+        return ifNotCached;
+    }
+
+    const _ref_Li9za2lrby5tanM_ = imports['./skiko.mjs'];
+    const _ref_QGpzLWpvZGEvY29yZQ_ = imports['@js-joda/core'];
+    
+    const js_code = {
+        'kotlin.captureStackTrace' : () => new Error().stack,
```

Comment:
I'm seeing a significant amount of samples in 'captureStackTrace' here.

The stack seems to have things like:
```
Wasm: Function <compose-benchmarks:benchmarks>.androidx.compose.runtime.LaunchedEffectImpl.onForgotten [baseline] 
Wasm: Function <compose-benchmarks:benchmarks>.androidx.compose.runtime.internal.PlatformOptimizedCancellationException.<init> [baseline] 
Wasm: Function <compose-benchmarks:benchmarks>.kotlin.coroutines.cancellation.CancellationException.<init> [baseline] 
```

Is this expected? It seems unfortunate for UI rendering to require stack trace collection to be fast.

Reply:
@eymar Why are we seeing these `CancellationException`s, is this expected?

Reply:
If `CancellationException`s are indeed used for "mostly regular" control-flow, that's fine, but does it make sense to materialize a stack trace for them? E.g., is there some option in `build.gradle.kts` that causes this, but is not usually enabled in production builds? Something along those lines? https://github.com/JetBrains/compose-multiplatform/blob/5eaf42c6f9af8fa36575501b50f42f81b32af523/benchmarks/multiplatform/benchmarks/build.gradle.kts#L51 (I tried commenting out that line, but that doesn't change anything wrt. `kotlin.captureStackTrace`, it's is still called >8000 times during the benchmark.)



In other words: Is this a realistic representation of what Kotlin/Wasm programs in the wild are doing? Are there plans to change this in the Kotlin toolchain?

Reply:
Hi everyone! 

Exceptions were changed in the recently released 2.2.20-Beta2. 
TL;DR:
* They read a stack value from a JS Error only when a stack is requested. 
* Each Kotlin Exception stores a reference to a JS Error
   * The JS Error has a back reference to the Kotlin Exception.
   * JS Error's name is modified after creation.
     * In 2.2.20 it will involve small data copying from wasm to js, since we don't use string builtins yet.
     * In 2.3+, it's going to be just writing a reference to JS String.

Reply:
@danleh 
>>The current behavior of eager reading a stack value cannot be configured.
>
>What does that mean? We need to update the toolchain (which I did) and the on-demand materialization/reading of the stack trace is going to be the new default?

Yes, it'll be a new default in 2.2.20-Beta2+.

>>In 2.3+, it's going to be just writing a reference to JS String.
>
>When will there be a stable 2.3 Kotlin/Wasm release?

2.2.20 is going to be released in the second half of September.
2.3.0 -- probably by the end of the year.